iothub/service/src/Amqp/ClientWebSocketTransport.cs

```diff
@@ -0,0 +1,412 @@
+    internal sealed class ClientWebSocketTransport : TransportBase, IDisposable
+    {
+        private static readonly AsyncCallback s_onReadComplete = OnReadComplete;
+        private static readonly AsyncCallback s_onWriteComplete = OnWriteComplete;
+        private static readonly TimeSpan s_closeTimeout = TimeSpan.FromSeconds(30);
+
+        private readonly EndPoint _localEndPoint;
+        private readonly EndPoint _remoteEndPoint;
+
+        // Disposables
+
+        private readonly ClientWebSocket _webSocket;
+        private readonly CancellationTokenSource _writeCancellationTokenSource;
+        private bool _isDisposed;
+
+        internal ClientWebSocketTransport(ClientWebSocket webSocket, EndPoint localEndpoint, EndPoint remoteEndpoint)
+            : base("clientwebsocket")
+        {
+            _webSocket = webSocket;
+            _localEndPoint = localEndpoint;
+            _remoteEndPoint = remoteEndpoint;
+            _writeCancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public override string LocalEndPoint => _localEndPoint.ToString();
+
+        public override string RemoteEndPoint => _remoteEndPoint.ToString();
+
+        public override bool RequiresCompleteFrames => true;
+
+        public override bool IsSecure => true;
+
+        public override void SetMonitor(ITransportMonitor usageMeter)
+        {
+            // Do Nothing
+        }
+
+        public override bool WriteAsync(TransportAsyncCallbackArgs args)
+        {
+            ThrowIfNotOpen();
+
+            Debug.Assert(args.Buffer != null || args.ByteBufferList != null, "must have a buffer to write");
+            Debug.Assert(args.CompletedCallback != null, "must have a valid callback");
+            args.Exception = null; // null out any exceptions
+
+            Task taskResult = WriteImplAsync(args);
+            if (WriteTaskDone(taskResult, args))
+            {
+                return false;
+            }
+
+            ToAsyncResult(taskResult, s_onWriteComplete, args);
+            return true;
+        }
+
+        public override bool ReadAsync(TransportAsyncCallbackArgs args)
+        {
+            ThrowIfNotOpen();
+
+            // Read with buffer list not supported
+            Debug.Assert(args.Buffer != null, "must have buffer to read");
+            Debug.Assert(args.CompletedCallback != null, "must have a valid callback");
+
+            Argument.ValidateBufferBounds(args.Buffer, args.Offset, args.Count);
+            args.Exception = null; // null out any exceptions
+
+            Task<int> taskResult = ReadImplAsync(args);
+            if (ReadTaskDone(taskResult, args))
+            {
+                return false;
+            }
+
+            ToAsyncResult(taskResult, s_onReadComplete, args);
+            return true;
+        }
+
+        public void Dispose()
+        {
+            if (!_isDisposed)
+            {
+                _webSocket.Abort();
+                _webSocket.Dispose();
```

Comment:
If you dispose the user's instance from options, they can't reopen the client, right?

Reply:
Right. I'll make it so users can opt out of that behavior

Reply:
So, if we put our own instance in options (if the user does not provide one) in the client ctor, we can free it optionally on dispose, rather than in this class, right?

Reply:
Why dispose the websocket at the service client level when this is the only level that uses it?

Reply:
Because it should last for the duration of the client despite how many times it is opened/closed?

Reply:
What if we re-use the web socket AMQP transport class (which holds this client web socket) for each open/close and only dispose it when the service client is disposed?

service client dispose -> amqp websocket layer dispose -> client web socket dispose

service client open -> create websocket layer if not already created

Reply:
If we do, then I'm fine with it.

However, it seems to me that the lifetime of the options class (and what it holds) is more congruent with the client than a single class in the transport layer.

Reply:
> service client dispose -> amqp websocket layer dispose -> client web socket dispose

But not if you disposed the one provided by the user, and then can't recreate it. Right?

Reply:
Right. We won't dispose the client web socket instance if the user provides it and opts out of us disposing it for them.